### DISCUSSION TITLE
A proper way to deal with export templates

### DISCUSSION THREAD

**Discussion by damiankaczkowski** [netbox-community/netbox#8065] [Q&A] [answered]
Hi.

I have a problem creating a proper "Export Templates" and need a little help.

Using netbox 3.1.1.

1. What is a generic way to get variable names that holds "obj" attributes?

I read https://demo.netbox.dev/static/docs/customization/export-templates/ and tried to export from "Custom Field" a "Content types" for which I used "content_types" field name (which is the name taken from import field name).

I used:
{% for customfield in queryset %}
{{ customfield.content_types }}
{% endfor %}

and instead of getting something like:
"dcim.device"
I got:
"contenttypes.ContentType.None"

Though that is a bug, but @jeremystretch pointed out in #8062 that it it not. I'm kind of surprised and I am wondering what is the best way to get the list of attributes that "obj" holds plus their values. The best would be to just craft some "generic" template and use it to export all data just to get a sneak peek of "obj" attributes and values so in the next step one would be able to make a proper, more detailed template. If this is impossible then what is the best way to get such info? API? Source code? Something else?

2. Why if I export for example "Tenants" using "All Data" build-in template I get different values than when I export via custom crafted Export Template?

For example:
 - Custom Export Template - gives me "None" value for "{{ tenant.group }}"
 - "All Data" build-in template - gives me an empty value for a "Group".

How to properly extract empty value from "{{ tenant.group }}" using "Export Template" when Group is not set ?

Many thanks in advance for help.
Greets.

**Comment by candlerb** [comment, marked as answer]
> I am wondering what is the best way to get the list of attributes that "obj" holds plus their values.

To explore what attributes are available on a given object, the [nbshell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/) is a useful tool.

```
cd /opt/netbox
. venv/bin/activate
python3 netbox/manage.py nbshell

>>> obj = Device.objects.first()
>>> dir(obj)
['DoesNotExist', 'Meta', 'MultipleObjectsReturned', '__class__', '__delattr__', ...
```

The source code for the model is also a good place to look, e.g. for [Device](https://github.com/netbox-community/netbox/blob/v3.1.1/netbox/dcim/models/devices.py#L437)

**Comment by damiankaczkowski** [comment]
Thank you @candlerb for the hint will look into that.

Regarding second question - @kkthxbye-code pointed out that for "{{ tenant.group }}" I should use "{{ tenant.group|default('', true) }}" which works in this case. Thank you again @kkthxbye-code .
